lib/notificationhub.go

```diff
@@ -0,0 +1,295 @@
+			continue
+		}
+	}
+
+	if _url.Scheme == schemeServiceBus || _url.Scheme == "" {
+		_url.Scheme = schemeDefault
+	}
+
+	_url.Path = hubPath
+	_url.RawQuery = url.Values{apiVersionParam: {apiVersionValue}}.Encode()
+	return &NotificationHub{
+		SasKeyName:  sasKeyName,
+		SasKeyValue: sasKeyValue,
+		HubURL:      _url,
+
+		client:                  utils.NewHubHTTPClient(),
+		expirationTimeGenerator: utils.NewExpirationTimeGenerator(),
+		regIDPath:               xmlpath.MustCompile("/entry/content/*/RegistrationId"),
+		eTagPath:                xmlpath.MustCompile("/entry/content/*/ETag"),
+		expTmPath:               xmlpath.MustCompile("/entry/content/*/ExpirationTime"),
+	}
+}
+
+// SetHTTPClient makes it possible to use a custom http client
+func (h *NotificationHub) SetHTTPClient(c utils.HTTPClient) {
+	h.client = c
+}
+
+// SetExpirationTimeGenerator makes is possible to use a custom generator
+func (h *NotificationHub) SetExpirationTimeGenerator(e utils.ExpirationTimeGenerator) {
+	h.expirationTimeGenerator = e
+}
+
+// Send publishes notification directly
+// Format tags according to https://docs.microsoft.com/en-us/azure/notification-hubs/notification-hubs-tags-segment-push-message
+// ex. "(follows_RedSox || follows_Cardinals) && location_Boston"
+// or nil if no tags should be used
+func (h *NotificationHub) Send(ctx context.Context, n *Notification, tags *string) ([]byte, error) {
+	b, err := h.send(ctx, n, tags, nil)
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.Send: %s", err)
+	}
+	return b, nil
+}
+
+// SendDirect publishes notification to a specific device
+func (h *NotificationHub) SendDirect(ctx context.Context, n *Notification, deviceHandle string) ([]byte, error) {
+	b, err := h.sendDirect(ctx, n, deviceHandle)
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.SendDirect: %s", err)
+	}
+	return b, nil
+}
+
+// Schedule pusblishes a scheduled notification
+// Format tags according to https://docs.microsoft.com/en-us/azure/notification-hubs/notification-hubs-tags-segment-push-message
+// or nil if no tags should be used
+func (h *NotificationHub) Schedule(ctx context.Context, n *Notification, tags *string, deliverTime time.Time) ([]byte, error) {
+	b, err := h.send(ctx, n, tags, &deliverTime)
+	if err != nil {
+		return nil, fmt.Errorf("notificationHub.Schedule: %s", err)
+	}
+
+	return b, nil
+}
+
+// Registration reads one specific registration
+func (h *NotificationHub) Registration(ctx context.Context, deviceID string) (*RegistrationResult, []byte, error) {
+	var (
+		result = &RegistrationResult{}
+		regURL = h.generateAPIURL("registrations")
+	)
+	regURL.Path = path.Join(regURL.Path, deviceID)
+	rawResponse, err := h.exec(ctx, getMethod, regURL, Headers{}, nil)
+	if err != nil {
+		return nil, rawResponse, err
+	}
+	if err = xml.Unmarshal(rawResponse, &result); err != nil {
+		return nil, rawResponse, err
+	}
+	result.RegistrationContent.normalize()
+	return result, rawResponse, nil
```

Comment:
nil, rawResponse, result for symmetry

Reply:
That would create a very different result